use errno wrapper instead of io::Error in Result types

acatangiu opened the pull request:
Because io::Error type can have an ErrorKind that is meaningless to a vmm, a simple errno wrapper is preferred. Occasionally, the io::Error function that converts an errno to ErrorKind has some surprising mappings, although many of them simply map to Other, so I wish to sidestep the problem entirely.

For display purposes, the Display impl can simply pass through to io::Error, which does have a useful Display impl for any relevant errno.

Fixes #35

acatangiu committed:
vcpu: use errno::Error wrapper instead of io::Error

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
vm: use errno::Error wrapper instead of io::Error

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
system: use errno::Error wrapper instead of io::Error

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu committed:
device: use errno::Error wrapper instead of io::Error

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

andreeaflorescu commented on src/ioctls/vcpu.rs:
```diff
@@ -8,13 +8,14 @@
 use kvm_bindings::{CpuId, Msrs};
 use kvm_ioctls::*;
+use vmm_sys_util::errno::Error as Errno;
+use vmm_sys_util::errno::Result as ErrnoResult;
```
I noticed that in the other commits you use errno::Error. I think it's better to be consistent here as well and import `vmm_sys_util::errno` instead.

acatangiu replied:
Oh, this is leftover code, good catch!

acatangiu committed:
fix some clippy warnings

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu commented:
@andreeaflorescu @sameo @jiangliu I've addressed the last issue listed in https://github.com/rust-vmm/kvm-ioctls/issues/59 so that we can release `kvm-ioctls v0.4.0`

jiangliu commented on Cargo.toml:
```diff
@@ -11,7 +11,7 @@ license = "Apache-2.0 OR MIT"
 libc = ">=0.2.39"
 kvm-bindings = { version = ">=0.2.0", features = ["fam-wrappers"] }
-vmm-sys-util = ">=0.1.1"
+vmm-sys-util = ">=0.2.1"
```
Should it be 0.2.0?

acatangiu replied:
0.2.1 contains a fix for clone of FamSafeWrappers which were released in 0.2.0, I chose to enforce using the fixed version of the crate.

andreeaflorescu commented on src/ioctls/mod.rs:
```diff
@@ -25,7 +24,7 @@ pub mod vm;
 /// This typedef is generally used to avoid writing out io::Error directly and
 /// is otherwise a direct mapping to Result.
-pub type Result<T> = result::Result<T, io::Error>;
+//pub type Result<T> = result::Result<T, io::Error>;
```
If we don't want to use this anymore, we should also delete the comment and the commented code.

acatangiu replied:
done

andreeaflorescu commented on src/lib.rs:
````diff
@@ -198,6 +198,6 @@ pub use ioctls::vm::{IoEventAddress, NoDatamatch, VmFd};
 /// ```
 /// #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-/// use kvm_ioctls::{KvmRunWrapper, Result};
+/// use kvm_ioctls::KvmRunWrapper;
````
Can you also add a comment in the commit message to say that the Result is no longer exported and the reason behind that?

acatangiu replied:
done

acatangiu committed:
completely switch from io::Error to errno::Error

Changed `kvm_ioctls::Result` to work with `errno::Result` instead
of `io::Result`.

No longer exporting `kvm_ioctls::Result`. Users of this crate should
not have to use `kvm_ioctls::Result` outside the crate.

Now exporting `kvm_ioctls::Error` type definition so that users of
this crate can create their own wrapping errors without having to know
the `Error` type used internally by this crate.

Signed-off-by: Adrian Catangiu <acatan@amazon.com>

acatangiu commented:
Changed `kvm_ioctls::Result` to work with `errno::Result` instead of `io::Result`.
    
No longer exporting `kvm_ioctls::Result`. Users of this crate should not have to use `kvm_ioctls::Result` outside the crate.
    
Now exporting `kvm_ioctls::Error` type definition so that users of this crate can create their own wrapping errors without having to know the `Error` type used internally by this crate.

```rust
enum UserError {
  FailedSomeOperation(kvm_ioctls::Error),
}
```
No need to assume `io::Error` or `errno::Error` used internally by `kvm_ioctls`.

@andreeaflorescu @jiangliu what do you think?

andreeaflorescu reviewed (approved):
As a future improvement we might want to have a typedef for KvmError or similar and replace errno::Error with KvmError. This should help if we need to change the error type in the future as we only need to change it in one place.
